### PR TITLE
Extended Settings

### PR DESCRIPTION
**Pull request by dzucconi**
Don't merge yet; just opening for now.

### CODE REVIEW

File: apps/settings/components/form_card/index.coffee [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      success: ->
+        $submit.text 'Success'
+        location.reload()
+
+      error: (xhr) ->
+        $errors
+          .show()
+          .html """
+            #{xhr.responseJSON.message}<br>
+            #{xhr.responseJSON.description}
```
Comment (dzucconi) [inline apps/settings/components/form_card/index.coffee]: Turns out we need to return better error messages from the server

File: apps/settings/components/form_card/index.coffee [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    serializer = new Serializer $el
+
+    $submit
+      .prop 'disabled', true
+      .text 'Saving...'
+
+    $.ajax
+      url: $el.data('action')
+      method: $el.data('method')
```
Comment (dzucconi) [inline apps/settings/components/form_card/index.coffee]: We're not doing anything useful with models here so... no models.

File: apps/settings/components/form_card/index.styl [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    > .bordered-select
+      margin 0
+
+  &__submit
+    margin unit 0 0 0
+
+    > button
+      border-top 1px solid bordergray
+      font-size s_middle
+      color textgray
```
Comment (dzucconi) [inline apps/settings/components/form_card/index.styl]: This is mostly throwaway code. Just trying to keep it contained.

File: apps/settings/components/tab_content/index.styl [hunk 4]
```diff
@@ -0,0 +1,5 @@
+.tab-content
+  margin unit auto
+
+  &__title
+    display none
```
Comment (dzucconi) [inline apps/settings/components/tab_content/index.styl]: Where I'll handle the mobile stuff. This should really have been part of the actual tab component but I'm not going to try to merge any modifications to it.

File: apps/settings/templates/tabs/billing.jade [hunk 5]
```diff
@@ -8,4 +8,28 @@ block content
...
+                  selected= user.get('plan') ? 'premium_monthly' : null
+                ) Premium $5 / month
+
+        .form-card__field
+          label Next billing date
+          label 1/1/2018
+
+      .form-card__submit
+        button.button--wide.button--border
+          | Save
```
Comment (dzucconi) [inline apps/settings/templates/tabs/billing.jade]: I'm going to go entirely through Stripe for this stuff. Our server should only concern itself with the bare minimum of determining who has what plan and their validity.

**Comment by dzucconi** [comment]
So, I know there are no specs. But do specs run as part of the build process?

File: apps/settings/components/form_card/index.coffee (hunk 1 above)
Comment (broskoski) [inline apps/settings/components/form_card/index.coffee]: Interesting.

File: apps/settings/components/form_card/index.coffee (hunk 2 above)
Comment (broskoski) [inline apps/settings/components/form_card/index.coffee]: Fine by me!

File: apps/settings/templates/tabs/billing.jade (hunk 5 above)
Comment (broskoski) [inline apps/settings/templates/tabs/billing.jade]: Yes, good plan.

**Comment by broskoski** [comment]
So specs would actually run as part of the build process, there just aren't any currently :|

Gonna look at this actual ❌

**Comment by broskoski** [comment]
Looks like you need to go down one more level for those stylesheet requires in the settings app.

File: lib/middleware/view_mode.coffee [hunk 6]
```diff
@@ -1,6 +1,6 @@
 # Set view mode from cookie
 #
 module.exports = (req, res, next) ->
-  mode = if (req.user?.get('is_pro') is true and req.cookies['view_mode']?) then req.cookies['view_mode'] else 'grid'
```
Comment (dzucconi) [inline lib/middleware/view_mode.coffee]: True is True

File: models/user.coffee [hunk 7]
```diff
@@ -1,45 +1,49 @@
...
-  href: -> "#{sd.APP_URL}/#{@get('slug')}"
+  href: ->
+    "#{APP_URL}/#{@get('slug')}"
 
   slugOrId: ->
-    @get('slug') || @id
+    @get('slug') or @id
 
   isPremium: ->
-    @get('is_pro') is true or @get('is_pro') is 'true'
```
Comment (dzucconi) [inline models/user.coffee]: True is True

**Comment by dzucconi** [comment]
All that's left now is

- [ ] Get it working on mobile
- [ ] Add avatar upload to the general tab (and fix it)
- [ ] Premium module should trigger form change
- [ ] Modalize it for the premium CTA (or just link to the billing page for now)

File: apps/settings/components/avatar_uploader/lib/file_type_regex.coffee [hunk 8]
```diff
@@ -0,0 +1 @@
+module.exports = /(\.|\/)(gif|jpe?g|png|ai|eps|kml|kmz|mb|ma|tex|texi|texinfo|tfm|fla|webm|ind|indd|key|pages|pdf|epub|psd|torrent|mp3|wav|aac|oga|ogg|wma|midi|aiff|mpeg|mpg|mpg4|mp4|mp4v|swa|swf|ttc|ttf|otf|pgp|numbers|fxp|latex|mov|avi|h264|ogv|docx|doc|ppt|pptx|xls|xlsx|xlt|tif|tiff|webloc)$/i
```
Comment (broskoski) [inline apps/settings/components/avatar_uploader/lib/file_type_regex.coffee]: All these files are ok? Don't we want to limit to images?

Comment (dzucconi) [inline apps/settings/components/avatar_uploader/lib/file_type_regex.coffee]: oh  yeah; that was just copypasta'd from what we have now. good catch

File: apps/settings/components/avatar_uploader/view.coffee [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    this
+
+  postRender: ->
+    @els =
+      form: @$('.js-upload')
+      submit: @$('.js-submit')
+
+    @els.form.fileupload
+      acceptFileTypes: fileTypeRegex
+      maxFileSize: 104857600 # 100MB # TODO: Extract
```
Comment (broskoski) [inline apps/settings/components/avatar_uploader/view.coffee]: Let's make this smaller for avatars (at least on the client)

File: apps/settings/components/payment_methods/index.jade [hunk 10]
```diff
@@ -0,0 +1,55 @@
+form
+  if customer.related().sources.length > 0
+    h3 Payment Methods
+
+    table.payment-methods__cards: tbody
+      each source in customer.related().sources.models
```
Comment (broskoski) [inline apps/settings/components/payment_methods/index.jade]: 👌

File: apps/settings/components/payment_methods/lib/styles.coffee [hunk 11]
```diff
@@ -0,0 +1,7 @@
+{ camelize } = require 'underscore.string'
+
+module.exports = ($el, properties) ->
+  properties.reduce (memo, property) ->
+    memo[camelize(property, true)] = $el.css(property)
+    memo
+  , {}
```
Comment (broskoski) [inline apps/settings/components/payment_methods/lib/styles.coffee]: 👁

File: apps/settings/components/avatar_uploader/view.coffee [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+      data: url: url
+
+    poll = () =>
+      @model.fetch().then =>
+        unless @model.hasChanged 'avatar_image'
+          setTimeout poll, 1000
+        else
+          @els.submit.text 'Avatar updated'
+
+    poll()
```
Comment (dzucconi) [inline apps/settings/components/avatar_uploader/view.coffee]: avatar

File: assets/settings.coffee [hunk 13]
```diff
@@ -1,3 +1,5 @@
+Backbone = require 'backbone'
 
 $ ->
-  require("../apps/auth/client/settings.coffee").init()
+  Backbone.$ = $
```
Comment (broskoski) [inline assets/settings.coffee]: :)

File: components/form/serializer.coffee [hunk 14]
```diff
@@ -0,0 +1,32 @@
+_ = require 'underscore'
+
+module.exports = class Serializer
```
Comment (broskoski) [inline components/form/serializer.coffee]: Hello

**Comment by broskoski** [comment]
That ❌  is weird, any ideas?

File: apps/settings/components/avatar_uploader/view.coffee (hunk 12 above)
Comment (broskoski) [inline apps/settings/components/avatar_uploader/view.coffee]: Nice, fwiw, we still send a pusher update for this. channel: "user_#{id}", action: "user-#{updated}". Poll-ing might be easier though.

Comment (dzucconi) [inline apps/settings/components/avatar_uploader/view.coffee]: I always assumed the current iteration never worked actually, since there was no feedback.

**Comment by dzucconi** [comment]
❌  is due to differing relative paths in the requires. For whatever reason. I guess I'll just not declare my CSS requires, like the fool I am.

**Comment by dzucconi** [comment]
(And I'll deal with the ❌  when I'm done)

**Comment by dzucconi** [comment]
I'm going to go ahead and merge this so I can take a look at staging.

**Pull request merged**